packages/odyssey/src/scss/components/_table.scss

```diff
@@ -239,7 +240,7 @@
       ~ tbody {
         tr:first-child {
           td:first-child {
```

Comment:
these selectors are really gnarly which means they are hard to understand and maintain in the future.

Do you think these are a side effect of not having JavaScript available in Odyssey? To mock up the different states we could insetad apply classes with JS and that would allow for flatter selectors and simpler CSS.

Reply:
Similar answer as above: since `table` is a specced element, these help enforce the use of proper markup + we can rely on them to always be present.  That is, it matters to the browser what order these elements are in and what they're nested under, so it doesn't bother me to reflect that in our styling. 

In this hyper-specific case of a variant table media query, it also allows us to treat locations and elements as names - because they will always occur in the same place, in the same way, in every table - without having to rely on a class name to communicate that, even when the alternate style isn't present.

e.g. everyone (who should be fiddling with sass anyway) knows exactly where the first `td` in a row is without needing to cross-check documentation on what the class `.table--linear-variant-heading` is for or where it shows up.

Reply:
@edburyenegren-okta Thanks I'll try and address this in parts in an effort to be more clear

> these help enforce the use of proper markup

I don't agree with using CSS as an enforcement mechanism for valid markup. In practice, developers will miss the subtle loss of styling that we are relying on for enforcement. If they do notice the problem, they'll have no visibility into why it is broken.

A better tool for that would be javascript, where we can noisily throw a helpful exception as soon as the violation occurs. Or even better, perform static analysis on our codebase to catch issues prior to run time.

> It matters to the browser what order these elements are in and what they're nested under, so it doesn't bother me to reflect that in our styling.

[See my response above for more of my thoughts on this](https://github.com/okta/odyssey/pull/534#discussion_r444334237)

I'm not certain sure I fully understand that latter half of your message, but my interpretation is that you're saying element and contextual selectors are more meaningful because we can rely on developers understanding how the markup is structured? Is this correct?

If I'm understanding correctly, I believe this gets back to the same discussion above. High specificity and complex DOM aware selectors will make our code base difficult to understand and refactor. Full stop! We won't ever need to document CSS classes (please god I hope so!) but when I see a class or two on an element, and I can search the styles to find those classes. I can immediately understand what styles are being applied. In fact, I have an easier time reasoning about this because I no longer need to consider the context of the element such as its general sibling position to a `<caption>` element, indirect descendent position from a `<tbody>` and further indirect descendant position from a `:first-child` `<tr>`, or the fact that I am the `:first-child` in a list as the `<td>`.

You're correct that table markup will not often change because the semantic elements are important, but we'll still need to provide variants in a way that doesn't need to fight upstream against increased specificity. Furthermore, it is nice if our code comes across in a single "voice". If we style with flat classes everywhere, new developers can ramp up more quickly and the project because easier to understand.

Reply:
@jeffbelser-okta thanks for bringing this up. Since these changes arent part of this PR, can you please log an issue and reference this thread? We can pick up the conversation there and determine a path forward!

Reply:
(From above, but consolidating.)

>If our selectors are always coupled closely to the structure of the DOM, or perhaps written in a way that provides different styles based on an awareness of their content, or perhaps relying on ever increasing specificity to apply the styles we want for a specific context than we will be cursing ourselves in the future. This will become difficult to maintain and understand.

No one's against this afaik. You'll notice that the rest of our codebase avoids relying on nested selectors as often as possible unless there is some true need for the distinction.

There are huge benefits to avoiding specificity race conditions - it's one of the biggest problems with Courage/Admin - but this is currently closer to a styled HTML element, not a custom component.  Creating classes like `.table--header`, `.table--row`, `.table--header-cell`, `.table--body`, `.table--body-row-heading`, `.table--first-child-cell-in-row` etc to cover each of the valid constructions of a table doesn't feel more clear as a maintainer than `thead th` or `tbody tr:first-child`. In this (incredibly limited) case, it's an obfuscation of a known, unchanging model.

Contrast this with our Input components: we do use adjacent sibling selectors there to allow for state detection between `label` and `input`, but every component part utilizes a class selector because the markup for constructing the entire molecule (fieldset, label, input, custom hint component, custom error messaging) has no standard.

In either case, I'm unclear on how a presence or lack of JS affects the selector choice here - the element selectors were intentional based on the model of `table`, not a lack of functionality in CSS or Sass. The alternative could be semantic class names replicating the semantics of table structure as you've suggested, but that's achievable in plain CSS.

----

@arnoldsandoval-okta all that said, the linear display for mobile tables was an experiment we intended to remove anyway. Could you yank this for now?